# TP NoSQL

Prénom :

Nom : 

Pour ce TP noté vous avez à trois heures pour le réaliser et 48 heures pour le finir si besoin. Le rendu final est attendu pour dimanche 29 septembre 2024 à 23 h 59.

Vous serrez évalués sur les requêtes mais également sur la qualité et la propreté du code. Pour chaque question, une réponse présentant les résultats est attendue.

In [1]:
# installation des packages pandas et pymongo
!pip install -r requirements.txt --quiet

In [2]:
import pymongo
import pandas

URI = 'mongodb+srv://mango_user:RRBEbgmVaiP8ooi3@cluster-but-sd.aomgg.mongodb.net/?retryWrites=true&w=majority&appName=cluster-but-sd'
client = pymongo.MongoClient(URI)
db = client.tp

# output the name of the collections in the database
print("Collections: ", db.list_collection_names())

Collections:  ['airbnb', 'restaurants']


1. Combien de logements sont la base de données ?

In [11]:
c = db.airbnb.count_documents({})
print(c)

5555


 Rep = Il y a 5555 logements dans la base de données

2. Quel est le prix moyen par ville ? Trier les villes par prix décroissant.

In [97]:
prix = db.airbnb.aggregate([
    { "$group": { "_id": "$address.market","average_price": { "$avg": "$price" }}},
    { "$sort": { "average_price": -1 }}
])
pandas.DataFrame(list(prix))

_id                        average_price
0               Hong Kong  762.4781906300484652665589660743134
1          Rio De Janeiro  525.8059701492537313432835820895522
2   Other (International)                               445.75
3                Istanbul  367.9454545454545454545454545454545
4                   Kauai  288.8208955223880597014925373134328
5                    Maui  286.5882352941176470588235294117647
6                    Oahu  212.2964426877470355731225296442688
7                  Sydney  197.7142857142857142857142857142857
8          The Big Island  179.4604316546762589928057553956835
9                New York  139.6260296540362438220757825370675
10       Other (Domestic)                               128.00
11                                                      115.50
12              Barcelona  100.9477848101265822784810126582278
13               Montreal  100.2330246913580246913580246913580
14                  Porto  69.12815884476534296028880866425993

Rep = La ville avec la moyenne de prix la plus élevée est Hong kong ($762), suivie de Rio de Janeiro ($252,8)...

3. Afficher la liste de tous les différents équipements qui existent.

In [69]:
equip = db.airbnb.aggregate([
    { "$unwind": "$amenities" },
    { "$group": { "_id": "$amenities" } },
    { "$project": { "équipments": "$_id", "_id": 0 } }
])

pandas.DataFrame(list(equip))

équipments
0                                 Ironing Board
1                                     Hot water
2                                       Hot tub
3                                   Stair gates
4                                       Bathtub
..                                          ...
181                             Host greets you
182                                Sun loungers
183  translation missing: en.hosting_amenity_50
184                       Wide clearance to bed
185                             Convection oven

[186 rows x 1 columns]

Rep = Il existe plusieurs équipements (Fer a repasser, eau chaude, baignoire...)

4. Combien de propriétés incluent le Wifi dans les équipements ?

In [61]:
wifi = db.airbnb.count_documents({"amenities": "Wifi"})
print(wifi)

5303


Rep = Il y a 5303 propriétés sur 5555 qui incluent le Wifi

5. Afficher le nom de tous les logements ainsi que le nombre de chambres et de lits qu'ils contiennent (ne pas afficher l'ID)

In [67]:
logement_tot = db.airbnb.aggregate([
       { "$group": { "_id": "$name", 
                 "nb_chambre":{"$sum": "$bedrooms"},
                 "nb_lits": { "$sum": "$beds" }}} 
])
pandas.DataFrame(list(logement_tot))

_id  nb_chambre  nb_lits
0     Ampla Suíte - Próx. a Barra da Tijuca,Jeunesse...           1        1
1            taksim meydanına yürüyerek sadece 3 dakika           2        3
2                                        Bohemian Suite           1        1
3                                  NEW APARTMENT OPORTO           1        2
4      Heart of Honolulu, 2BD gem! Free Garage Parking!           2        2
...                                                 ...         ...      ...
5533                         DOUBLE ROOM for 1 or 2 ppl           1        1
5534                                    SUNNY RAMBLAS 2           2        4
5535  Sea Views & Porto's City Park - Ocean Windows ...           3        4
5536         Single Sweaty Room 2 stop from Grand Pazar           1        1
5537          HOMEnFUN Sants close to the Train station           3        4

[5538 rows x 3 columns]

In [ ]:
Rep = Le logement Ampla Suíte contient 1 chambres et 1 lits, le logement taksim meydanına yürüyerek sadece 3 dakika  contient 2 chambres et 3 lits

6. Afficher le nom et le prix des logements situés à Porto.

In [59]:
Porto = db.airbnb.find({"address.market" : "Porto"},
            { "_id": 0, "name" : 1 ,"price": 1 }     )
pandas.DataFrame(list(Porto))

name   price
0                Ribeira Charming Duplex   80.00
1                      Be Happy in Porto   30.00
2    Downtown Oporto Inn (room cleaning)   40.00
3                         FloresRooms 3T   31.00
4     A Casa Alegre é um apartamento T1.   40.00
..                                   ...     ...
549         Number283 - Your home abroad   48.00
550        Cozy apartment downtown Porto   58.00
551                   Soares Guest House   90.00
552   Central, Modern, University, Metro   41.00
553     StayIN Oporto Música - Apartment  500.00

[554 rows x 2 columns]

A Porto la Ribeira Charming Duplex 	coutent 80$, la StayIN Oporto Música - Apartment quant à elle coutent 500$

7. Quels sont les 5 hôtes les plus populaires (ceux dont les propriétés ont reçu le plus de commentaires) ?

In [112]:
hotes = db.airbnb.aggregate([ 
    { "$unwind": "$reviews" },
    { "$group": {"_id": "$name", "review_count": {"$sum": 1}}},
     { "$sort": { "review_count": -1 }},
     { "$limit": 5 }
])
pandas.DataFrame(list(hotes))

_id  review_count
0                    #Private Studio - Waikiki Dream           533
1  Near Airport private room, 2 bedroom granny fl...           469
2           La Sagrada Familia (and metro) 4 blocks!           463
3                PRIVATE Room in Spacious, Quiet Apt           420
4                      traditional and Charming room           408

les 5 hotes les plus poplaires sont : #Private Studio - Waikiki Dream avec 533 commentaires, Near Airport private room, 2 bedroom granny avec 469 commentaires, La Sagrada Familia (and metro) 4 blocks! avec 463 commentaires, PRIVATE Room in Spacious, Quiet Apt avec 420 commentaires, et traditional and Charming room avec 408 commentaires.

8. Quelles sont les 6 villes ayant le plus de logements disponibles à la location ?

In [77]:
six = db.airbnb.aggregate([ 
    { "$group": {"_id": "$address.market", "count": {"$sum": 1}}},
     { "$sort": {"count": -1}},
     { "$limit": 6 }
])
pandas.DataFrame(list(six))

_id  count
0   Istanbul    660
1   Montreal    648
2  Barcelona    632
3  Hong Kong    619
4     Sydney    609
5   New York    607

In [ ]:
Istanbul compte 660 logements disponibles, Montreal compte 648 logements, Barcelone compte 632 logements...

9. Combien de propriétés acceptent plus de 4 invités et ont une caution de moins de 300€ ?

In [114]:
caution = db.airbnb.aggregate([ 
    {"$match": {"accommodates": {"$gt": 4}, "security_deposit": {"$lt": 300}}},
     {"$project": {
            "name": 1,
            "security_deposit": 1,
            "accommodates": 1,
            "room_type": 1}},
])
pandas.DataFrame(list(caution))

_id                                               name  \
0    10006546                            Ribeira Charming Duplex   
1    10051164                 Catete's Colonial Big Hause Room B   
2     1022200                Kailua-Kona, Kona Coast II 2b condo   
3    10133350                          2 bedroom Upper east side   
4    10166883   Large railroad style 3 bedroom apt in Manhattan!   
..        ...                                                ...   
446   9567840                  Large Comfortable CBD Convenience   
447   9990304     March Madness Special! Aina Nalu Platinum D107   
448   9372696               Economy Family Room for 4-5pax (306)   
449   9612902  1921 15kms from Sydney CBD with Pool Spa parking.   
450   9917859                   StayIN Oporto Música - Apartment   

           room_type  accommodates security_deposit  
0    Entire home/apt             8           200.00  
1       Private room             8             0.00  
2    Entire home/apt             6           200.00  
3    Entire home/apt             5             0.00  
4    Entire home/apt             9           199.00  
..               ...           ...              ...  
446  Entire home/apt             6             0.00  
447  Entire home/apt             6           200.00  
448     Private room             5             0.00  
449  Entire home/apt             9             0.00  
450  Entire home/apt            16           200.00  

[451 rows x 5 columns]

Le Ribeira Charming Duplex accepete 8 personnes avec une caution de 200$, le Catete's Colonial Big Hause Room B accepte 8 personnes avec une caution de 0$...

10. Donner les 20 utilisateurs qui ont fait le plus de commentaires (afficher seulement l'ID et le nom de l'utilisateur).

In [138]:
vingts= db.airbnb.aggregate([
    { "$unwind": "$reviews" },
    { "$group": {"_id": "$reviews.reviewer_id", "util": { "$first": "$reviews.reviewer_name" },"com": { "$sum": 1 }}},
    { "$sort": { "com": -1 } },
    { "$limit": 20 },
    { "$project": { "_id": 1, "util": 1} }
])

pandas.DataFrame(list(vingts))

_id      util
0    20775242    Filipe
1    67084875      Nick
2     2961855       Uge
3   162027327     Thien
4    20991911      Lisa
5     1705870     David
6    60816198      Todd
7    12679057      Jodi
8    55241576  Courtney
9    69140895      Lisa
10   78093968     David
11   47303133     Lance
12   57325457      Mary
13   24667379     Karen
14   86665925     Chris
15   25715809     Megan
16   73708321   Gonzalo
17  128210181   Branden
18   61469899      Erik
19   34005800       Dan

Filipe qui a l'id: 20775242 fait partie des 20 personnes qui ont fait le plus de commentaires, Nick qui a l'id: 67084875 fait partie des 20 personnes qui ont fait le plus de commentaires...

11. Parmi les logements à Sydney, quel est la note moyenne des visiteurs ?

In [126]:
syndney = db.airbnb.aggregate([
    { "$match": { "address.market": "Sydney" } },
    { "$group": { "_id":"$address.market" ,"avg": { "$avg": "$review_scores.review_scores_rating" }}},
])

pandas.DataFrame(list(syndney))

_id       avg
0  Sydney  93.45567

La note moyenne des visiteurs des logements à Sydney est de 93.45567.

12. (bonus) Afficher les logements qui contiennent le mot "park" dans leur nom

In [180]:
park= db.airbnb.find({ "name" : { "$regex":" park ", "$options": 'm' }}, 
                      { "_id" : 0, "name" : 1})
pandas.DataFrame(list(park))

name
0   Close to park and sea: Light and airy family apt
1  Terrace- park view, great location, quiet street.
2            2 bedroom + 1 sofa, free car park place

Il y a 47 logements qui contiennent le mots "Park"

13. (bonus) Afficher le nom des logements ayant une lattitude comprise entre 36,1 et 40,6 dont le prix est entre 100 et 200 euros.

In [142]:
lattitude = db.airbnb.aggregate([
     { "$match": { "address.location.coordinates.1": { "$gte": 36.1, "$lte": 40.6 },"price": { "$gte": 100, "$lte": 200 }}},
    { "$project": { "nom_du_logement": "$name", "_id": 0,}}
])
pandas.DataFrame(list(lattitude))

nom_du_logement
0              Cozy Comfy Suite of Rockaway Beach
1       Coney Island MCU Park Wi fi Cable Apt****
2  Sheepshead Bay, 1-bdr apartment. Close to all!

3 logements ont une lattitude comprise entre  36,1 et 40,6 et ont un prix entre 100 et 200 euros: 
- Cozy Comfy Suite of Rockaway Beach
- Coney Island MCU Park Wi fi Cable Apt****
- Sheepshead Bay, 1-bdr apartment. Close to all!